httpcore/_async/connection_pool.py

```diff
@@ -238,28 +238,27 @@ def _assign_requests_to_connections(self) -> List[AsyncConnectionInterface]:
         those connections to be handled seperately.
         """
         closing_connections = []
-        idling_connections = {c for c in self._connections if c.is_idle()}
+        idling_count = sum(1 for c in self._connections if c.is_idle())
 
         # First we handle cleaning up any connections that are closed,
         # have expired their keep-alive, or surplus idle connections.
         for connection in list(self._connections):
             if connection.is_closed():
                 # log: "removing closed connection"
                 self._connections.remove(connection)
-                idling_connections.discard(connection)
+                idling_count -= int(connection.is_idle())
             elif connection.has_expired():
                 # log: "closing expired connection"
                 self._connections.remove(connection)
-                idling_connections.discard(connection)
                 closing_connections.append(connection)
+                idling_count -= int(connection.is_idle())
             elif (
-                connection.is_idle()
-                and len(idling_connections) > self._max_keepalive_connections
+                connection.is_idle() and idling_count > self._max_keepalive_connections
             ):
                 # log: "closing idle connection"
                 self._connections.remove(connection)
-                idling_connections.discard(connection)
                 closing_connections.append(connection)
+                idling_count -= 1
```

Comment:
Just for myself interest: Could we have *single-loop incremental* idling count instead of *double-loop decremental*?
```python
        closing_connections = []
        idling_count = 0

        # First we handle cleaning up any connections that are closed,
        # have expired their keep-alive, or surplus idle connections.
        for connection in list(self._connections):
            if connection.is_closed():
                # log: "removing closed connection"
                self._connections.remove(connection)
            elif connection.has_expired():
                # log: "closing expired connection"
                self._connections.remove(connection)
                closing_connections.append(connection)
            elif connection.is_idle():
                if idling_count > self._max_keepalive_connections:
                    # log: "closing idle connection"
                    self._connections.remove(connection)
                    closing_connections.append(connection)
                else:
                    idling_count += 1

```

Reply:
We could simplify even more if we recreate the `self._connections` and not `remove` from it:
```python
connections = []
closing_connections = []
idling_count = 0

# First we handle cleaning up any connections that are closed,
# have expired their keep-alive, or surplus idle connections.
for connection in self._connections:
    if connection.is_closed():
        # log: "removing closed connection"
        pass
    elif connection.has_expired():
        # log: "closing expired connection"
        closing_connections.append(connection)
    elif connection.is_idle():
        if idling_count > self._max_keepalive_connections:
            # log: "closing idle connection"
            closing_connections.append(connection)
        else:
            idling_count += 1
            connections.append(connection)

self._connections = connections
```

Reply:
I also concern about this new change, because `idling_count` will became _dynamic_ through iteration while previously it was _static_ out of iteration. However, we can test it.


> ```python
> self._connections = connections
> ```

I prefer to avoid re-assigning at here 🤷‍♂️

Reply:
Ok, I [did test](https://github.com/T-256/httpcore/commit/07c066074e8a5a87eaf39334170cd57bd817cb06) and failed:
https://github.com/T-256/httpcore/actions/runs/9482448452/job/26127347871

Reply:
Good catch, thanks!

Reply:
And fixed by [this change](https://github.com/T-256/httpcore/commit/ffb789a4f692a8aafaea7a15abc5e4648295465f). now tests passing :)

Reply:
Thanks thats just great, I applied your diff!